src/external/matchers/deepeval.ts

```diff
@@ -0,0 +1,125 @@
+// These metrics are ported from DeepEval.
+// https://docs.confident-ai.com/docs/metrics-conversation-relevancy. See APACHE_LICENSE for license.
+import dedent from 'dedent';
+import { getAndCheckProvider, fail } from '../../matchers';
+import { getDefaultProviders } from '../../providers/defaults';
+import type { GradingConfig, GradingResult } from '../../types';
+import invariant from '../../util/invariant';
+import { getNunjucksEngine } from '../../util/templates';
+
+const nunjucks = getNunjucksEngine(undefined, false, true);
+
+export interface Message {
+  input: string;
+  output: string | object;
+}
+
+function fromVars(vars?: Record<string, string | object>) {
+  if (!vars) {
+    return {};
+  }
+
+  const ret: Record<string, string> = {};
+  for (const [key, value] of Object.entries(vars)) {
+    if (typeof value === 'object') {
+      ret[key] = JSON.stringify(value, null, 2);
+    } else {
+      ret[key] = JSON.stringify(value, null, 2).slice(1, -1);
+    }
+  }
+
+  return ret;
+}
+
+function parseJsonResponse(output: string): any {
+  const match = output.match(/```(?:json)?([^`]+)```/);
+  if (match) {
+    output = match[1].trim();
+  }
+
+  try {
+    return JSON.parse(output);
+  } catch (err) {
+    throw new Error(`Failed to parse LLM output (${output}) as JSON: ${err}`);
+  }
+}
+
+export async function matchesConversationRelevance(
+  messages: Message[],
+  threshold: number,
+  vars?: Record<string, string | object>,
+  grading?: GradingConfig,
+): Promise<Omit<GradingResult, 'assertion'>> {
+  const textProvider = await getAndCheckProvider(
+    'text',
+    grading?.provider,
+    (await getDefaultProviders()).gradingProvider,
+    'conversation relevancy check',
+  );
+
+  const tokensUsed = {
```

Comment:
you can simplify this. Look at how you're using tokens in other places. Also please type this you will notice fields are missing